java/src/hu/holyoil/controller/AIController.java

```diff
@@ -102,14 +105,119 @@ public void RemoveTeleportGate(TeleportGate teleportGate)  {
+                            return;
+                        }
+                    }
+                }
+            }//if all fails continue and hope for best
+            //don't drill if water or uranium would react to the Sun, otherwise drill if makes sense
+            if (current.GetLayerCount() > 0 && !(current.GetIsNearbySun() && current.GetLayerCount() == 1
+                    && current.GetResource()!=null)) {
+                robot.Drill();
+                Logger.Return();
+                return;
+            }
+            Random random = new Random();
+            Asteroid target;
+            int chosenIndex = random.nextInt(neighbouringAsteroids.size());
+            int start = chosenIndex;
+            boolean listOver = false;
+            boolean shouldMove = false;
+            //does NOT account for player movement
+            do {
+                if (chosenIndex == neighbouringAsteroids.size() - 1) {
+                    chosenIndex = -1; //goes through the list but starts at a random index, goes from end to beginning here
+                }
+                chosenIndex++;
 
-        robot.Move(robot.GetOnAsteroid().GetRandomNeighbour());
+                target = neighbouringAsteroids.get(chosenIndex);
+                if (target.GetLayerCount() > 0 && target.GetResource() != null
+                        && !(target.GetIsNearbySun() && target.GetLayerCount() == 1)) {
+                    //look for neighbour that: has layers, isn't near Sun AND would possibly react, isn't empty
+                    shouldMove = true;
+                }
 
-        // todo: proper intelligence
+                if (chosenIndex == start) {
+                    //if back at beginning, exit cycle
+                    listOver = true;
+                }
+            } while (!listOver && !shouldMove);
+            start = chosenIndex;
+            listOver = false;
+            while (!listOver && !shouldMove) {//only enters if didn't find suitable neighbour
+
+                if (chosenIndex == neighbouringAsteroids.size() - 1) {
+                    chosenIndex = -1;
+                }
+                chosenIndex++;
+
+                target = neighbouringAsteroids.get(chosenIndex);
+                if (target.GetLayerCount() > 0) {
+                    //find a drillable empty neighbour
+                    shouldMove = true;
+                }
+
+                if (chosenIndex == start) {
+                    listOver = true;
+                }
+            }
+            if (shouldMove) {//found desirable neighbour
+                robot.Move(neighbouringAsteroids.get(chosenIndex));
+            } else {
+                if (tpAvailable)//go through tp, hope for richer asteroids
+                    robot.Move(current.GetTeleporter());
+                else//everything failed, all neighbours suck, doesn't matter where robot moves
+                    robot.Move(current.GetRandomNeighbour());
```

Comment:
If everything failed, should it move? Mmint az is lehetőség, hogy ne mozogjon, nem kötelező.
Ha így is ok, akkor nem mondom, hogy muszáj lenne átírni ezt a kódrészletet, maradhat így is.

Reply:
De mozogjon, mert ha ott marad a kovi korben se lesznek jo szomszedjai